contracts/TradeTypesLib2.sol

```diff
@@ -0,0 +1,134 @@
+/*
+ * Hubii Nahmii
+ *
+ * Compliant with the Hubii Nahmii specification v0.12.
+ *
+ * Copyright (C) 2017-2018 Hubii AS
+ */
+
+pragma solidity >=0.4.25 <0.6.0;
+
+import {MonetaryTypesLib} from "./MonetaryTypesLib.sol";
+import {NahmiiTypesLib} from "./NahmiiTypesLib.sol";
+
+/**
+ * @title     TradeTypesLib
+ * @dev       Data types centered around trade
+ */
+library TradeTypesLib2 {
+    //
+    // Enums
+    // -----------------------------------------------------------------------------------------------------------------
+    enum CurrencyRole {Intended, Conjugate}
+    enum LiquidityRole {Maker, Taker}
+    enum Intention {Buy, Sell}
+    enum TradePartyRole {Buyer, Seller}
+
+    //
+    // Structures
+    // -----------------------------------------------------------------------------------------------------------------
+    struct Operator {
+        uint256 id;
+        string data;
+    }
+
+    struct OrderPlacement {
+        Intention intention;
+
+        int256 amount;
+        NahmiiTypesLib.IntendedConjugateCurrency currencies;
+        int256 rate;
+
+        NahmiiTypesLib.CurrentPreviousInt256 residuals;
+    }
+
+    struct OrderParty {
+        uint256 nonce;
+        address wallet;
+
+        NahmiiTypesLib.IntendedConjugateCurrency balances;
+
+        string data;
+    }
+
+    struct Order {
+        OrderParty party;
+
+        OrderPlacement placement;
+
+        NahmiiTypesLib.WalletOperatorSeal seals;
+        uint256 blockNumber;
+
+        Operator operator;
+    }
+
+    struct OrderCancellation {
+        OrderParty party;
+
+        NahmiiTypesLib.WalletOperatorSeal seals;
+        uint256 blockNumber;
+
+        Operator operator;
+    }
+
+    struct TradeOrder {
+        int256 amount;
+        NahmiiTypesLib.WalletOperatorHashes hashes;
+        NahmiiTypesLib.CurrentPreviousInt256 residuals;
+    }
+
+    struct TradeParty {
+        uint256 nonce;
+        address wallet;
+
+        uint256 rollingVolume;
+
+        LiquidityRole liquidityRole;
+
+        TradeOrder order;
```

Comment:
Residuals will only be recorded in the trades and not issued as an updated order?

Reply:
Well, we should define this. Does the partial filling of an order issue a new order receipt in addition to the trade receipt? If true then residuals have a place in `Order`. Otherwise I don't think so.

Reply:
If the contracts are able to resolve challenges using residuals on a trade, that would be fine from the perspective of the backend I think. Internally we will have to do some form of book keeping anyway to track residuals on orders, but we dont want to expose that information unless the SCs needs it.

Reply:
The contracts are indeed able to cope with residuals only listed on trade receipts and not on order receipts.